src/Shared/TagWriter/TagWriter.cs

```diff
@@ -0,0 +1,301 @@
+                }
+
+                break;
+
+            // All other types are converted to strings including the following
+            // built-in value types:
+            // case nint:    Pointer type.
+            // case nuint:   Pointer type.
+            // case ulong:   May throw an exception on overflow.
+            // case decimal: Converting to double produces rounding errors.
+            default:
+                try
+                {
+                    var stringValue = TruncateString(Convert.ToString(tag.Value, CultureInfo.InvariantCulture), tagValueMaxLength);
+                    if (stringValue == null)
+                    {
+                        return this.LogUnsupportedTagTypeAndReturnFalse(tag.Key, tag.Value);
+                    }
+
+                    this.WriteStringTag(state, tag.Key, stringValue);
+                }
+                catch
+                {
+                    // If ToString throws an exception then the tag is ignored.
+                    return this.LogUnsupportedTagTypeAndReturnFalse(tag.Key, tag.Value);
+                }
+
+                break;
+        }
+
+        return true;
+    }
+
+    protected abstract void WriteIntegralTag(TTagState state, string key, long value);
+
+    protected abstract void WriteFloatingPointTag(TTagState state, string key, double value);
+
+    protected abstract void WriteBooleanTag(TTagState state, string key, bool value);
+
+    protected abstract void WriteStringTag(TTagState state, string key, string value);
+
+    protected abstract void WriteArrayTag(TTagState state, string key, TArrayState value);
+
+    protected abstract void OnUnsupportedTagDropped(
+        string tagKey,
+        string tagValueTypeFullName);
+
+    [return: NotNullIfNotNull(nameof(value))]
+    private static string? TruncateString(string? value, int? maxLength)
+    {
+        return maxLength.HasValue && value?.Length > maxLength
+            ? value.Substring(0, maxLength.Value)
+            : value;
+    }
+
+    private void WriteArrayTagInternal(TTagState state, string key, Array array, int? tagValueMaxLength)
+    {
+        var arrayState = this.arrayWriter.BeginWriteArray();
+
+        // This switch ensures the values of the resultant array-valued tag are of the same type.
+        switch (array)
+        {
+            case char[] charArray: this.WriteToArray(arrayState, charArray); break;
+            case string[]: this.ConvertToStringArrayThenWriteArrayTag(arrayState, array, tagValueMaxLength); break;
+            case bool[] boolArray: this.WriteToArray(arrayState, boolArray); break;
+            case byte[] byteArray: this.WriteToArray(arrayState, byteArray); break;
+            case sbyte[] sbyteArray: this.WriteToArray(arrayState, sbyteArray); break;
+            case short[] shortArray: this.WriteToArray(arrayState, shortArray); break;
+            case ushort[] ushortArray: this.WriteToArray(arrayState, ushortArray); break;
+#if NETFRAMEWORK
```

Comment:
Instead of repeating `#if NETFRAMEWORK` we could get int[], unint[], long[] under one block.

Reply:
Updated

Reply:
Woops yes sorry about that I just put it back.